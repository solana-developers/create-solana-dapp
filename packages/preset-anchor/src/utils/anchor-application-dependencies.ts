import { addDependenciesToPackageJson, Tree } from '@nx/devkit'
import { packageVersion } from '@solana-developers/preset-common'

export function anchorApplicationDependencies(tree: Tree) {
  return addDependenciesToPackageJson(
    tree,
    {
      '@coral-xyz/anchor': packageVersion['@coral-xyz'].anchor,
    },
    {},
  )
}
