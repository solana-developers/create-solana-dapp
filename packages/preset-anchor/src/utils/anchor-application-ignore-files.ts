import { Tree } from '@nx/devkit'
import { updateIgnoreEntries } from '@solana-developers/preset-common'

export function anchorApplicationIgnoreFiles(tree: Tree, projectRoot: string) {
  const ignoreFiles = {
    git: [
      '.anchor',
      `${projectRoot}/target/deploy`,
      `${projectRoot}/target/release`,
      `${projectRoot}/target/sbf-solana-solana`,
      `${projectRoot}/target/.rustc_info.json`,
      `!${projectRoot}/target/idl/*.json`,
      `!${projectRoot}/target/types/*.ts`,
      'node_modules',
      'test-ledger',
      '.yarn',
    ],
    prettier: ['.anchor', 'target', 'node_modules', 'dist', 'tmp', 'build', 'test-ledger'],
  }

  updateIgnoreEntries(tree, '.gitignore', ignoreFiles.git)
  updateIgnoreEntries(tree, '.prettierignore', ignoreFiles.prettier)
}
