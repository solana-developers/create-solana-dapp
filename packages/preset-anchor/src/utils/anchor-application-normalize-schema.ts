import { AnchorApplicationSchema } from '../'

export function anchorApplicationNormalizeSchema(options: AnchorApplicationSchema): AnchorApplicationNormalizedSchema {
  return {
    ...options,
    skipFormat: options.skipFormat ?? false,
    template: options.template ?? 'counter',
  }
}

export type AnchorApplicationNormalizedSchema = Required<AnchorApplicationSchema>
export type AnchorApplicationTemplate = AnchorApplicationSchema['template']
