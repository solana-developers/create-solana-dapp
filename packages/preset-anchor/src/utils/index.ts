export * from './anchor-application-ignore-files'
export * from './anchor-application-dependencies'
export * from './anchor-application-normalize-schema'
