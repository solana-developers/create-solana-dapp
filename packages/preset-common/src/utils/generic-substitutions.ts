import { names } from '@nx/devkit'

export interface GenericSubstitutions {
  name: string
  npmScope: string
  anchor: string
  anchorName: string
  preset?: 'next' | 'react'
}
export function genericSubstitutions({ anchor, anchorName, name, npmScope, preset }: GenericSubstitutions) {
  return {
    ...names(name),
    anchor: names(anchor),
    anchorName: names(anchorName),
    npmScope,
    currentFullYear: new Date().getFullYear(),
    preset: preset ?? 'react',
  }
}
