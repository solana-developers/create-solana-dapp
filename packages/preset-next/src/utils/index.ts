export * from './next-application-substitutions'
export * from './generate-next-application'
export * from './normalize-next-application-schema'
