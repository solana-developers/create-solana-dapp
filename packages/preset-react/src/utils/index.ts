export * from './application-tailwind-config'
export * from './generate-react-application'
export * from './get-ui-dependencies'
export * from './normalize-react-application-schema'
export * from './react-application-dependencies'
export * from './react-application-run-scripts'
export * from './react-application-substitutions'
export * from './wallet-adapter-dependencies'
