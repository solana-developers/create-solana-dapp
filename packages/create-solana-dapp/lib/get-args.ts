import { intro, log } from '@clack/prompts'
import { program } from 'commander'
import { PackageManager } from 'nx/src/utils/package-manager'
import { anchorTemplates } from './anchor-templates'
import { getAppInfo } from './get-app-info'
import { GetArgsResult } from './get-args-result'
import { getPresets } from './get-presets'
import { getPrompts, getUiLibraries } from './get-prompts'

export async function getArgs(argv: string[], pm: PackageManager = 'npm'): Promise<GetArgsResult> {
  // Get app info from package.json
  const app = getAppInfo()

  // Get info about the presets
  const { presets, presetValues } = getPresets(app.version)

  // Get the result from the command line
  const input = program
    .name(app.name)
    .version(app.version, '-V, --version', help('Output the version number\n'))
    .argument('[name]', 'Name of the project (default: <prompt>)')
    .option(
      '-p, --preset <preset>',
      help(`Preset to use (default: <prompt>, options: ${presetValues.join(', ')})\n`),
      (value: string) => {
        if (!presetValues.includes(value)) {
          throw new Error(`Invalid preset: ${value}`)
        }
        return value
      },
    )
    .option('--ui <ui-library>', help(`UI library to use (default: <prompt>)\n`))
    .option(
      '-a, --anchor <template>',
      help(`Name of the Anchor template to use (default: <prompt>, set to "none" to prevent adding Anchor)\n`),
      (value: string) => {
        if (!anchorTemplates.includes(value)) {
          throw new Error(`Invalid anchor template: ${value}`)
        }
        return value
      },
    )
    .option('--anchor-build', help(`Build the anchor project`), false)
    .option('--anchor-name <name>', help(`Anchor project name (default: anchor)\n`))
    .option('--web-name <name>', help(`Web project name (default: web)\n`))
    .option('--web-port <port>', help(`Web project port (default: 3000)\n`))
    .option('-pm, --package-manager <package-manager>', help(`Package manager to use (default: npm)\n`))
    .option('--yarn', help(`Use yarn as the package manager`), false)
    .option('--pnpm', help(`Use pnpm as the package manager`), false)
    .option('-d, --dry-run', help('Dry run (default: false)\n'))
    .helpOption('-h, --help', help('Display help for command\n'))
    .addHelpText(
      'after',
      `
Examples:
  $ ${app.name} my-app --preset react
  $ ${app.name} my-app --preset react --package-manager yarn
  $ ${app.name} my-app --preset react --anchor hello-world
      `,
    )
    .parse(argv)

  // Get the optional name argument (positional)
  const name = input.args[0]

  // Get the options from the command line
  const result = input.opts()

  let packageManager = result.packageManager ?? pm

  // The 'yarn' and 'pnpm' options are mutually exclusive, and will override the 'packageManager' option
  if (result.pnpm && result.yarn) {
    log.error(`Both pnpm and yarn were specified. Please specify only one.`)
    throw new Error(`Both pnpm and yarn were specified. Please specify only one.`)
  }
  if (result.pnpm) {
    packageManager = 'pnpm'
  }
  if (result.yarn) {
    packageManager = 'yarn'
  }

  // Take the result from the command line and use it to populate the options
  const options: GetArgsResult = {
    anchor: result.anchor,
    anchorBuild: result.anchorBuild,
    anchorName: result.anchorName ?? 'anchor',
    dryRun: result.dryRun ?? false,
    name: name ?? '',
    package: '',
    packageManager,
    pnpm: result.pnpm ? result.pnpm : false,
    preset: result.preset,
    ui: result.ui,
    yarn: result.yarn ? result.yarn : false,
    webName: result.webName ?? 'web',
    webPort: result.webPort ?? 3000,
  }

  // Display the intro
  intro(`${app.name} ${app.version}`)

  // Get the prompts for any missing options
  const prompts = await getPrompts({ options, presets })

  // Populate the options with the prompts
  if (prompts.name) {
    options.name = prompts.name
  }
  if (prompts.preset) {
    options.preset = prompts.preset
  }
  if (prompts.ui) {
    options.ui = prompts.ui as string
  }
  if (prompts.anchor) {
    options.anchor = prompts.anchor
  }

  // Validate the options
  if (options.preset && !presetValues.includes(options.preset)) {
    log.error(`Invalid preset: ${options.preset}`)
    throw new Error(`Invalid preset: ${options.preset}`)
  }

  if (!anchorTemplates.includes(options.anchor ?? '')) {
    log.error(`Invalid anchor template: ${options.anchor}`)
    throw new Error(`Invalid anchor template: ${options.anchor}`)
  }

  const libs = getUiLibraries(options.preset as 'next' | 'react')
  if (options.ui && !libs.map((l) => l.value).includes(options.ui)) {
    log.error(`Invalid ui library for preset ${options.preset}: ${options.ui}`)
    throw new Error(`Invalid ui library for preset ${options.preset}: ${options.ui}`)
  }

  // Create the package string
  // TODO: In the future, we should allow the user to specify the package and version using the preset option
  if (options.package === '') {
    const packageName = presets.find((preset) => preset.value === options.preset)?.package
    const packageVersion = presets.find((preset) => preset.value === options.preset)?.version
    options.package = `${packageName}@${packageVersion}`
  }

  return options
}

// Helper function to add a newline before the text
function help(text: string) {
  return `

  ${text}`
}
