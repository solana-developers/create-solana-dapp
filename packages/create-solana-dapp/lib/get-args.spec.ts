import { PackageManager } from 'nx/src/utils/package-manager'
import { getArgs } from './get-args'

describe('get-args', () => {
  describe('expected args', () => {
    it('should get args with minimal parameters', async () => {
      const result = await getArgs(['', '', 'my-app', '--preset', 'next', '--ui', 'tailwind', '--anchor', 'counter'])
      expect(result.packageManager).toBe('npm')
      expect(result).toMatchSnapshot()
    })

    it.each(['npm', 'yarn', 'pnpm'])(
      `should get args with minimal parameters and package manager %s`,
      async (pm: PackageManager) => {
        const result = await getArgs(
          ['', '', 'my-app', '--preset', 'next', '--ui', 'tailwind', '--anchor', 'counter'],
          pm,
        )
        expect(result).toMatchSnapshot()
      },
    )

    it.each([
      '--preset next --ui tailwind --anchor counter',
      '--preset react --ui tailwind --anchor counter',
      '--preset react --anchor none --ui none',
      '--preset react --anchor counter --anchor-name program --ui tailwind',
      '--preset=react --anchor=counter --anchor-name=program --ui=tailwind',
      '--preset=react --anchor=counter --anchor-name=program --anchor-build --ui=tailwind',
      '--preset react --anchor none --ui none --pnpm',
      '--preset react --anchor none --ui none --yarn',
      '--preset react --anchor none --ui none --web-port=4200',
    ])('should get args with preset: %s', async (preset: string) => {
      const result = await getArgs(['', '', 'my-app', ...preset.split(' ')])

      expect(result).toMatchSnapshot()
    })
  })

  describe('unexpected args', () => {
    it('should fail with invalid preset', async () => {
      expect.assertions(1)
      try {
        await getArgs(['', '', 'my-app', '--preset', 'invalid'])
      } catch (e) {
        expect(e.message).toContain('Invalid preset: invalid')
      }
    })

    it('should fail with invalid anchor template', async () => {
      expect.assertions(1)
      try {
        await getArgs(['', '', 'my-app', '--preset', 'react', '--anchor', 'invalid'])
      } catch (e) {
        expect(e.message).toContain('Invalid anchor template: invalid')
      }
    })

    it('should fail with invalid ui library', async () => {
      expect.assertions(1)
      try {
        const x = await getArgs(['', '', 'my-app', '--preset', 'react', '--ui', 'jquery-ui'])
        console.log(x)
      } catch (e) {
        expect(e.message).toContain('Invalid ui library for preset react: jquery-ui')
      }
    })

    it('should fail with two package managers', async () => {
      expect.assertions(1)
      try {
        const x = await getArgs([
          '',
          '',
          'my-app',
          '--preset',
          'react',
          '--ui',
          'none',
          '--anchor',
          'none',
          '--yarn',
          '--pnpm',
        ])
        console.log(x)
      } catch (e) {
        expect(e.message).toContain('Both pnpm and yarn were specified. Please specify only one.')
      }
    })
  })
})
