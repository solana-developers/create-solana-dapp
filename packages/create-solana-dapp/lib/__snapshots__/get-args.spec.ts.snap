// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-args expected args should get args with minimal parameters 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "next",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with minimal parameters and package manager npm 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "next",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with minimal parameters and package manager pnpm 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "pnpm",
  "pnpm": false,
  "preset": "next",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with minimal parameters and package manager yarn 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "yarn",
  "pnpm": false,
  "preset": "next",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset next --ui tailwind --anchor counter 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-next@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "next",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor counter --anchor-name program --ui tailwind 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "program",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "react",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor none --ui none --pnpm 1`] = `
{
  "anchor": "none",
  "anchorBuild": false,
  "anchorName": "program",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "pnpm",
  "pnpm": true,
  "preset": "react",
  "ui": "none",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor none --ui none --web-port=4200 1`] = `
{
  "anchor": "none",
  "anchorBuild": false,
  "anchorName": "program",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "react",
  "ui": "none",
  "webName": "web",
  "webPort": "4200",
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor none --ui none --yarn 1`] = `
{
  "anchor": "none",
  "anchorBuild": false,
  "anchorName": "program",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "yarn",
  "pnpm": false,
  "preset": "react",
  "ui": "none",
  "webName": "web",
  "webPort": 3000,
  "yarn": true,
}
`;

exports[`get-args expected args should get args with preset: --preset react --anchor none --ui none 1`] = `
{
  "anchor": "none",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "react",
  "ui": "none",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset react --ui tailwind --anchor counter 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "anchor",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "react",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset=react --anchor=counter --anchor-name=program --anchor-build --ui=tailwind 1`] = `
{
  "anchor": "counter",
  "anchorBuild": true,
  "anchorName": "program",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "react",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;

exports[`get-args expected args should get args with preset: --preset=react --anchor=counter --anchor-name=program --ui=tailwind 1`] = `
{
  "anchor": "counter",
  "anchorBuild": false,
  "anchorName": "program",
  "dryRun": false,
  "name": "my-app",
  "package": "@solana-developers/preset-react@0.0.1",
  "packageManager": "npm",
  "pnpm": false,
  "preset": "react",
  "ui": "tailwind",
  "webName": "web",
  "webPort": 3000,
  "yarn": false,
}
`;
